File: app/src/main/java/cm/aptoide/pt/account/view/ImageValidator.java
package cm.aptoide.pt.account.view;

import cm.aptoide.pt.account.view.exception.InvalidImageException;
import java.util.ArrayList;
import java.util.List;
import rx.Completable;
import rx.Scheduler;

public class ImageValidator {
  private static final int MIN_IMAGE_HEIGHT = 600;
  private static final int MAX_IMAGE_HEIGHT = 10240;
  private static final int MIN_IMAGE_WIDTH = 600;
  private static final int MAX_IMAGE_WIDTH = 10240;
  private static final int MAX_IMAGE_SIZE_IN_BYTES = 5242880;
  private final Scheduler scheduler;
  private final ImageInfoProvider imageInfoProvider;

  public ImageValidator(Scheduler scheduler, ImageInfoProvider imageInfoProvider) {
    this.scheduler = scheduler;
    this.imageInfoProvider = imageInfoProvider;
  }

  /**
   * @return {@link InvalidImageException} if image is not valid.
   */
  public Completable validateOrGetException(String imagePath) {
    return Completable.defer(() -> {
      try {
        validate(imagePath);
        return Completable.complete();
      } catch (InvalidImageException ex) {
        return Completable.error(ex);
      }
    })
        .subscribeOn(scheduler);
  }

  private ImageInfo getInfo(String imagePath) throws InvalidImageException {
    ImageInfo info = imageInfoProvider.getInfo(imagePath);
    if (info == null) {
      final List<InvalidImageException.ImageError> errors = new ArrayList<>();
      errors.add(InvalidImageException.ImageError.ERROR_DECODING);
      throw new InvalidImageException(errors);
    }
    return info;
  }

  private void validate(String imagePath) throws InvalidImageException {
    ImageInfo imageInfo = getInfo(imagePath);
    final List<InvalidImageException.ImageError> errors = new ArrayList<>();
    if (imageInfo == null) {
      errors.add(InvalidImageException.ImageError.ERROR_DECODING);
    } else {
      if (imageInfo.getHeight() < MIN_IMAGE_HEIGHT) {
        errors.add(InvalidImageException.ImageError.MIN_HEIGHT);
      }
      if (imageInfo.getWidth() < MIN_IMAGE_WIDTH) {
        errors.add(InvalidImageException.ImageError.MIN_WIDTH);
      }
      if (imageInfo.getHeight() > MAX_IMAGE_HEIGHT) {
        errors.add(InvalidImageException.ImageError.MAX_HEIGHT);
      }
      if (imageInfo.getWidth() > MAX_IMAGE_WIDTH) {
        errors.add(InvalidImageException.ImageError.MAX_WIDTH);
      }
      if (imageInfo.getSize() > MAX_IMAGE_SIZE_IN_BYTES) {
        errors.add(InvalidImageException.ImageError.MAX_IMAGE_SIZE);
      }
    }

    if (!errors.isEmpty()) {
      throw new InvalidImageException(errors);
    }
  }
}

File: app/src/main/java/cm/aptoide/pt/account/view/ImageValidator.kt
package cm.aptoide.pt.account.view

import cm.aptoide.pt.account.view.exception.InvalidImageException
import cm.aptoide.pt.account.view.exception.InvalidImageException.ImageError
import rx.Completable
import rx.Scheduler
import java.util.*

open class ImageValidator(private val scheduler: Scheduler,
                          private val imageInfoProvider: ImageInfoProvider) {

  open fun validateOrGetException(imagePath: String): Completable? {
    return Completable.defer {
      validate(imagePath)
      return@defer Completable.complete()
    }
        .subscribeOn(scheduler)
  }

  @Throws(InvalidImageException::class)
  private fun getInfo(imagePath: String): ImageInfo {
    val info = imageInfoProvider.getInfo(imagePath)
    if (info == null) {
      val errors: MutableList<ImageError> = ArrayList()
      errors.add(ImageError.ERROR_DECODING)
      throw InvalidImageException(errors)
    }
    return info
  }

  @Throws(InvalidImageException::class)
  private fun validate(imagePath: String) {
    val imageInfo = getInfo(imagePath)
    val errors: MutableList<ImageError> = ArrayList()

    if (imageInfo.height < MIN_IMAGE_HEIGHT) {
      errors.add(ImageError.MIN_HEIGHT)
    } else if (imageInfo.height > MAX_IMAGE_HEIGHT) {
      errors.add(ImageError.MAX_HEIGHT)
    }

    if (imageInfo.width < MIN_IMAGE_WIDTH) {
      errors.add(ImageError.MIN_WIDTH)
    } else if (imageInfo.width > MAX_IMAGE_WIDTH) {
      errors.add(ImageError.MAX_WIDTH)
    }

    if (imageInfo.size > MAX_IMAGE_SIZE_IN_BYTES) {
      errors.add(ImageError.MAX_IMAGE_SIZE)
    }

    if (!errors.isEmpty()) {
      throw InvalidImageException(errors)
    }
  }

  companion object {
    private const val MIN_IMAGE_HEIGHT = 600
    private const val MAX_IMAGE_HEIGHT = 10240
    private const val MIN_IMAGE_WIDTH = 600
    private const val MAX_IMAGE_WIDTH = 10240
    private const val MAX_IMAGE_SIZE_IN_BYTES = 5242880
  }

}